### ISSUE TITLE
ECS agent 1.4.0 not reclaiming devicemapper space

### ISSUE THREAD

**Issue by rferreira** [aws/amazon-ecs-agent#772]
Howdy folks, I ran into an odd issue lately, here are the details:
* AWS ECS optimized image ami-48f9a52e
* AWS Agent 1.4.0

After a while the instance started throwing IO errors, running containers started to fail and new ones couldn't be started. 
```
[7019518.616037] Buffer I/O error on device dm-4, logical block 2710489
[7019518.621178] Buffer I/O error on device dm-4, logical block 2710490
[7019518.625887] Buffer I/O error on device dm-4, logical block 2710491
[7019518.630613] Buffer I/O error on device dm-4, logical block 2710492
[7019518.635383] Buffer I/O error on device dm-4, logical block 2710493
[7019518.640377] Buffer I/O error on device dm-4, logical block 2710494
[7019518.645137] Buffer I/O error on device dm-4, logical block 2710495
[7019520.892144] EXT4-fs warning (device dm-5): ext4_end_bio:329: I/O error -28 writing to inode 1055008 (offset 20918272 size 40960 starting block 2548722)
```

Now, for the kicker, there weren't images, containers or volumes consuming meaningful disk space:
```
[ec2-user@ip-10-1-7-80 ~]$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[ec2-user@ip-10-1-7-80 ~]$ docker images
REPOSITORY                                               TAG                 IMAGE ID            CREATED             SIZE
zzzzzzzz.dkr.ecr.us-east-1.amazonaws.com/zzzzzzz   v2.1.208            59ac0ed8e970        2 weeks ago         1.258 GB
amazon/amazon-ecs-agent                                  latest              90be67243de1        3 months ago        8.276 MB
# docker volume ls
DRIVER              VOLUME NAME

```
 which leads me to believe that the device mapper volumes were just not being reclaimed, is this a known issue by any chance?

**Comment by aaithal** [comment]
Hi @rferreira, you might be running into [this issue](https://github.com/moby/moby/issues/18867). Can you please follow remediation listed [there](https://github.com/moby/moby/issues/18867#issuecomment-166945185) to see if that fixes the issue in any way?

Thanks,
Anirudh

**Comment by cavis** [comment]
Also seeing my ECS instances running out of storage, since upgrading my AMIs from `ami-d69c74c0` to `ami-275ffe31` (ECS-agent version went from 1.14.0 to 1.14.1).

Have a cron running fstrim regularly (as documented [here in your developer guide](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/troubleshooting.html)), but it's not having any impact since upgrading.  Haven't been able to manually free up unused data blocks either.

Any suggestions?  Thanks.

**Comment by samuelkarp** [comment]
@cavis Can you provide the output of `docker info`, `docker ps -a`, `sudo lvs`, `sudo pvs`, and `sudo vgs`?  It'd also be good to look at the Docker daemon logs (in `/var/log/docker`) and the ECS agent logs (in `/var/log/ecs`).  You can collect all this information by running the [ECS log collector](https://github.com/awslabs/ecs-logs-collector).  If you're not comfortable providing this information publicly, please feel free to email it to me at skarp at amazon.com.

**Comment by cavis** [comment]
@samuelkarp I rolled our ECS AMI back to 1.14.0 ... but over the past 2 weeks it has also been gradually running out of space.

On the host system:

```
$ docker info | grep Space
 Data Space Used: 75.91 GB
 Data Space Total: 76.45 GB
 Data Space Available: 541.1 MB
 Metadata Space Used: 4.243 MB
 Metadata Space Total: 79.69 MB
 Metadata Space Available: 75.45 MB
 Thin Pool Minimum Free Space: 7.645 GB

$ lsblk
NAME                                                                                         MAJ:MIN   RM  SIZE RO TYPE MOUNTPOINT
xvda                                                                                         202:0      0    8G  0 disk
└─xvda1                                                                                      202:1      0    8G  0 part /
xvdcz                                                                                        202:26368  0   72G  0 disk
└─xvdcz1                                                                                     202:26369  0   72G  0 part
  ├─docker-docker--pool_tmeta                                                                253:0      0   76M  0 lvm
  │ └─docker-docker--pool                                                                    253:2      0 71.2G  0 lvm
  │   ├─docker-202:1-263238-7a4e23527d09524c99e36a7e1063b015cee03f3f8ca15060605643d9a68a03ac 253:3      0  100G  0 dm
  │   └─docker-202:1-263238-c5d72079d2d526abeb2707fc1d067cd34c485cf1c9728816221a83bdbfa7a411 253:4      0  100G  0 dm
  └─docker-docker--pool_tdata                                                                253:1      0 71.2G  0 lvm
    └─docker-docker--pool                                                                    253:2      0 71.2G  0 lvm
      ├─docker-202:1-263238-7a4e23527d09524c99e36a7e1063b015cee03f3f8ca15060605643d9a68a03ac 253:3      0  100G  0 dm
      └─docker-202:1-263238-c5d72079d2d526abeb2707fc1d067cd34c485cf1c9728816221a83bdbfa7a411 253:4      0  100G  0 dm
```

And in the only container (other than the `amazon/amazon-ecs-agent:latest`):

```
$ du -sh /
10.1G	/

$ df
Filesystem           1K-blocks      Used Available Use% Mounted on
/dev/mapper/docker-202:1-263238-c5d72079d2d526abeb2707fc1d067cd34c485cf1c9728816221a83bdbfa7a411
                     104804352  73435336  31369016  70% /
tmpfs                  2024928         0   2024928   0% /dev
tmpfs                  2024928         0   2024928   0% /sys/fs/cgroup
/dev/xvda1             8123812   1101580   6921984  14% /etc/resolv.conf
/dev/xvda1             8123812   1101580   6921984  14% /etc/hostname
/dev/xvda1             8123812   1101580   6921984  14% /etc/hosts
shm                      65536         0     65536   0% /dev/shm
tmpfs                  2024928         0   2024928   0% /proc/kcore
tmpfs                  2024928         0   2024928   0% /proc/latency_stats
tmpfs                  2024928         0   2024928   0% /proc/timer_list
tmpfs                  2024928         0   2024928   0% /proc/timer_stats
tmpfs                  2024928         0   2024928   0% /proc/sched_debug 
```

I have a cron running fstrim as root on the host system, which is failing to reclaim any space.  Running fstrim manually isn't working either:

```
*/5 * * * * docker ps -qa | xargs docker inspect --format="{{ .State.Pid }}" | xargs -IZ /sbin/fstrim /proc/Z/root/
```

Here are [all those logs](https://gist.github.com/cavis/91a766fe840878e3ca7c59cd9030f18a).  Only interesting thing I see in there is the "Error trying v2 registry: failed to register layer: devmapper" in the docker logs.  But that could be just a symptom of the system running out of space.

Also may be worth noting that I'm doing this in the userdata of my launch config, to get a larger disk size:

```
#!/bin/sh
echo 'EXTRA_DOCKER_STORAGE_OPTIONS="\${EXTRA_DOCKER_STORAGE_OPTIONS} --storage-opt dm.basesize=100G"' >> /etc/sysconfig/docker-storage-setup
```

**Comment by rferreira** [comment]
interesting potential tidbit, I'l also adjusting the dm.basesize via a user data script.

**Comment by samuelkarp** [comment]
@cavis Thanks for providing all that information.  It looks like your instance is configured to [use xfs](https://gist.github.com/cavis/91a766fe840878e3ca7c59cd9030f18a#file-docker-info-L11) as the backing filesystem for devicemapper.  We've seen problems with xfs being unable to free space once space has been exhausted (including issues where xfs would cause Docker to hang, see https://github.com/moby/moby/issues/20707) and we do not recommend using xfs.  I'm not sure which AMI you're using, but we've been configuring the ECS-optimized AMI to use ext4 since 2016.03.a.  Can you try running an instance with ext4 instead of xfs and see if that resolves your problem?  The most recent ECS-optimized AMI can be found [here](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-optimized_AMI_launch_latest.html).

@rferreira Did the suggestions that @aaithal provided help resolve your issue or are you still running into problems?  If you're still experiencing issues, can you run the [ECS logs collector](https://github.com/awslabs/ecs-logs-collector) and provide the logs?  If you're not comfortable posting those publicly, you can email them directly to me at skarp at amazon.com.

**Comment by cavis** [comment]
@samuelkarp How do you control what is used for the logical volume?  I'm using a launch config on the official ECS AMI `ami-d69c74c0` (2016.09), with storage setup like this:

![image](https://cloud.githubusercontent.com/assets/1410587/26407670/03eb1876-4059-11e7-9d92-641d5addd947.png)

**Comment by cavis** [comment]
Ah, okay, i see what i did.  I have a multipart userdata script with a double-escaped `$`.  So I end up with this as the last 2 lines of `/etc/sysconfig/docker-storage-setup`:

```
EXTRA_DOCKER_STORAGE_OPTIONS="--storage-opt dm.fs=ext4"
EXTRA_DOCKER_STORAGE_OPTIONS="\${EXTRA_DOCKER_STORAGE_OPTIONS} --storage-opt dm.basesize=100G"
```

So the ext4 option doesn't actually make it into the final `EXTRA_DOCKER_STORAGE_OPTIONS`.  My bad!  Will update that, and confirm that space is getting reclaimed.

**Comment by rferreira** [comment]
I'll give it a shot tomorrow and report back @samuelkarp

**Comment by seankovacs** [comment]
I'm running into an issue not with ECS agent but with a docker setup on EB where space isn't reclaimed  (writes and deletes from /tmp) and eventually running out of space, throwing the file system into read only mode. 

Here is a docker info:
`Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.6
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: 
 Metadata file: 
 Data Space Used: 5.166 GB
 Data Space Total: 12.73 GB
 Data Space Available: 7.568 GB
 Metadata Space Used: 786.4 kB
 Metadata Space Total: 16.78 MB
 Metadata Space Available: 15.99 MB
 Thin Pool Minimum Free Space: 1.273 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 4.9.27-14.31.amzn1.x86_64
Operating System: Amazon Linux AMI 2017.03
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.306 GiB
Name: ip-172-31-27-228
ID: CSZB:BELP:J7IT:EKZA:IBXS:GWM2:QE2M:OUOJ:TJGO:EGBR:AHNA:BVQV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/`


sudo lvs shows the data % going up, eventually going to 100%
` LV          VG     Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  docker-pool docker twi-aot--- 11.86g             42.53  4.79       `

**Comment by samuelkarp** [comment]
@rferreira Were you able to try out @aaithal's suggestions?

@seankovacs I'm not as familiar with the Elastic Beanstalk AMIs and what configuration they have.  Can you let me know the specific AMI ID you're using so I can take a look at it?

> where space isn't reclaimed (writes and deletes from /tmp) and eventually running out of space

@seankovacs This sounds exactly like what @aaithal described here: https://github.com/aws/amazon-ecs-agent/issues/772#issuecomment-297154618.  Can you try his suggestions?

**Comment by seankovacs** [comment]
@samuelkarp  I did end up  creating a cron job that runs every minute to fstrim the volume. Kinda annoying.

**Comment by rferreira** [comment]
hey folks, apologies for the slow response. Unfortunately, this issue hasn't resurfaced since I created this original ticket so I couldn't justify spending too much time on it. However, I did look into it today and, as noted this thread https://github.com/moby/moby/issues/18867, my issue is not reclaiming space of a running container (in which case fstrim makes total sense) but, instead, that deleted containers/images were not getting properly cleaned out of the dm pool - I would venture to say that this is an issue with the docker dm driver. If/when the issue resurfaces I'll gather the logs @samuelkarp suggested and post it to this thread. In the meantime, feel free to close this ticket.
